File: source/kalk/dati/implement/frazione.cpp
#include "dati/define/frazione.h"

frazione::frazione(int a, int b){
    if(b==0){
        throw std::string("il denominatore non puo essre 0");
    }
    if(b<0){
        a*=-1;
        b*=-1;
    }
    num = a;
    den = b;
    name="frazione";
}

frazione* frazione::clone() const{  return new frazione(*this);}

int frazione::getNum() const {    return num;}

void frazione::setNum(int a){    num=a;}

int frazione::getDen() const {     return den;}

void frazione::setDen(int a){    den=a;}

std::string frazione::getName()const{
    return name;
}

void frazione::setName(std::string s){
    name=s;
}

std::string frazione::toString()const{
    std::string s = std::to_string(num)+'/'+std::to_string(den);
    return s;
}

int frazione::MCD(int a, int b){ return b == 0  ? a : MCD(b, a % b);}

int frazione::getMCD() const{
    if(num<0)
        return frazione::MCD(-num,den);
    return frazione::MCD(num,den);
}

void frazione::reduce(){
    int a=getMCD();
    num=num/a;
    den=den/a;
}

frazione*  frazione::minima() const{
    frazione* a= new frazione(*this);
    a->reduce();
    return a;
}

double frazione::razionale() const{   return num/(double)den;}

frazione*  frazione::complementare() const{
    if(razionale()>1)
        throw std::string("frazione>1 non puo avere complementare");
    return new frazione(den-num, den);

}

frazione*  frazione::operator +(const frazione* b) const{
    long int n = (long int)num*b->den + (long int)b->num*den;
    long int d = (long int)den*b->den;
    if(n>INT_MAX ||n<INT_MIN ||  d>INT_MAX ||  d<INT_MIN){
        throw std::string("la somma non rispetta i limiti di INT");
    }
    frazione * a=new frazione(n,d);
    a->reduce();
    return a;
}

frazione*  frazione::operator -(const frazione*  b) const{
    long int n=(long int)num*b->den - (long int)b->num*den;
    long int d=(long int)den*b->den;
    if(n>INT_MAX ||n<INT_MIN ||  d> INT_MAX ||  d<INT_MIN)
        throw std::string("la differenza non rispetta i limiti di INT");
    frazione * a=new frazione(n, d);
    a->reduce();
    return a;
}

frazione*  frazione::operator *(const frazione*  b) const{
    long int n=(long int)num*b->num;
    long int d=(long int)den*b->den;
    if(n>INT_MAX ||n<INT_MIN ||  d> INT_MAX ||  d<INT_MIN){
        throw std::string("la moltiplicazione non rispetta i limiti di INT");
    }
    frazione * a=new frazione(n, d);
    a->reduce();
    return a;
}

frazione*  frazione::operator /(const frazione*  b) const{
    if(num==0)
        return new frazione();
    if(b->num==0)
        throw std::string("non si puo dividere per zero");
    long int n=(long int)num*b->den;
    long int d=(long int)den*b->num;
    if(n>INT_MAX ||n<INT_MIN ||  d> INT_MAX ||  d<INT_MIN){
        throw std::string("la divisione non rispetta i limiti di INT");
    }
    frazione * a=new frazione(n, d);
    a->reduce();
    return a;
}

bool frazione::operator==(const frazione*  a) const{    return razionale()==a->razionale();}

bool frazione::operator!=(const frazione*  a) const{    return razionale()!=a->razionale();}

bool frazione::operator<(const frazione*  a) const{    return razionale()<a->razionale();}

bool frazione::operator>(const frazione*  a) const{     return razionale()>a->razionale();}

File: source/java/frazione.java
import java.util.*;

public class frazione{
    private int num;
    private int den;
    private String name;
    public frazione(int n, int d){
        if(d<0){
            n=n*(-1);
            d=d*(-1);
        }
        num=n;
        den=d;
        name="frazione";
    }
    public void setNum(int i){
        num=i;
    }
    public int getNum(){
        return num;
    }
    public void setDen(int i){
        den=i;
    }
    public int getDen(){
        return den;
    }
    public void setName(String s){
        name=s;
    }
    public String getName(){
        return name;
    }
    public frazione clone(){
        return new frazione(num, den);
    }
    public String toString(){
        return String.valueOf(getNum())+"/"+String.valueOf(getDen());
    }
    public static int MCD(int x, int y){
        return y==0 ? x: MCD(y, x%y);
    }
    public int getMCD(){
        if (num<0){
            return MCD(-num,den);
        }
        return MCD(num,den);
    }
    public void reduce(){
        int a=getMCD();
        num=num/a;
        den=den/a;
    }
    public frazione minima(){
        frazione a=clone();
        a.reduce();
        return a;
    }
    public Double razionale(){
        double a=num;
        double b=den;
        return a/b;
    }
    public frazione complementare()throws alert{
        if(razionale()>1){
            throw new alert("frazione>1 non puo avere complementare");
        }
        return new frazione(getNum()-getDen(), getDen());
    }
    public frazione somma(frazione f)throws alert{
        Long n=Long.valueOf(num)*Long.valueOf(f.den)+Long.valueOf(f.num)*Long.valueOf(den);
        Long d=Long.valueOf(den)*Long.valueOf(f.den);
        if(n>Integer.MAX_VALUE || n<Integer.MIN_VALUE || d>Integer.MAX_VALUE || d<Integer.MIN_VALUE )
            throw new alert("la somma non rispetta i limiti di INT");
        frazione a= new frazione(n.intValue(),d.intValue());
        a.reduce();
        return a;
    }
    public frazione differenza(frazione f)throws alert{
        Long n=Long.valueOf(num)*Long.valueOf(f.den)-Long.valueOf(f.num)*Long.valueOf(den);
        Long d=Long.valueOf(den)*Long.valueOf(f.den);
        if(n>Integer.MAX_VALUE || n<Integer.MIN_VALUE || d>Integer.MAX_VALUE || d<Integer.MIN_VALUE )
            throw new alert("la differenza non rispetta i limiti di INT");
        frazione a= new frazione(n.intValue(),d.intValue());
        a.reduce();
        return a;
    }
    public frazione moltiplicazione(frazione f)throws alert{
        Long n=Long.valueOf(num)*Long.valueOf(f.num);
        Long d=Long.valueOf(den)*Long.valueOf(f.den);
        if(n>Integer.MAX_VALUE || n<Integer.MIN_VALUE || d>Integer.MAX_VALUE || d<Integer.MIN_VALUE )
            throw new alert("la moltiplicazione non rispetta i limiti di INT");
        frazione a= new frazione(n.intValue(),d.intValue());
        a.reduce();
        return a;
    }
    public frazione divisione(frazione f)throws alert{
        if(num==0){
            return new frazione(0, 1);
        }
        if(f.num==0){
            throw new alert("non si puo dividere per zero");
        }
        Long n=Long.valueOf(num)*Long.valueOf(f.den);
        Long d=Long.valueOf(den)*Long.valueOf(f.num);
        if(n>Integer.MAX_VALUE || n<Integer.MIN_VALUE || d>Integer.MAX_VALUE || d<Integer.MIN_VALUE )
            throw new alert("la divisione non rispetta i limiti di INT");
        frazione a= new frazione(n.intValue(),d.intValue());
        a.reduce();
        return a;
    }
    public boolean ugualianza(frazione f)throws alert{
        return razionale()==f.razionale();
    }
    public boolean diverso(frazione f)throws alert{
        return razionale()!=f.razionale();
    }
    public boolean maggiore(frazione f)throws alert{
        return razionale()>f.razionale();
    }
    public boolean minore(frazione f)throws alert{
        return razionale()<f.razionale();
    }
}